### DISCUSSION TITLE
Way to avoid repeating revset subexpressions

### DISCUSSION THREAD

**Discussion by pokey** [arxanas/git-branchless#562] [Ideas]
In [this gist](https://gist.github.com/pokey/7e1d21f3d3c3f0a700c3b07ef1aff20e#file-create-prs-for-stack-sh-L26-L27), I have to repeat a fairly verbose subexpression, and avoid it by defining the expression as a bash variable.  It would be good if there were some way to avoid needing to do this.  I have a couple ideas for how to avoid this problem

**Comment by pokey** [comment]
Define variables, as per https://github.com/arxanas/git-branchless/discussions/553#discussioncomment-3714647.  This would give:

```
git query "all_ancestors = ancestors(parents(\"$branch\")) & branches() & draft(); all_ancestors - ancestors(parents(all_ancestors))"
```

**Comment by pokey** [comment]
Support a syntax that is more like [jq](https://stedolan.github.io/jq/).  We could eg replace their `|` with `>`, and their `.` with `input()`.  So for example, it would be:

```
git query "ancestors(parents(\"$branch\")) & branches() & draft() > input() - ancestors(parents(input()))"
```

**Reply by arxanas** [reply]
Personally, I'm in favor of a pipe expression. It's rather common when writing a query interactively that you want to do more stuff that you just thought of, so a postfix variable binding syntax would be useful.

Currently, both `@` (from Git) and `.` (from Mercurial) mean `HEAD`. We could change `.` to mean "the result of the LHS of the pipe expression" with a default of `HEAD` when there is no pipe expression, which would work equivalently for all queries at present.

For the actual operator, I'm partial to `|>`, which is used in a few languages.

Proper variable binding would be useful if we get to the point of writing full scripts. The `.` syntax could be used for `git query` as well to indicate the input.

**Comment by pokey** [comment]
Support piping a list of commits into `query` and referring to it using eg `input()`.  Then we'd have:

```
git query -r "ancestors(parents(\"$branch\")) & branches() & draft()" | git query "input() - ancestors(parents(input()))"
```
